src/wallet/wallet.cpp

```diff
@@ -1868,8 +1868,18 @@ bool CWalletTx::IsTrusted() const
         if (parent == NULL)
             return false;
         const CTxOut& parentOut = parent->tx->vout[txin.prevout.n];
-        if (pwallet->IsMine(parentOut) != ISMINE_SPENDABLE)
-            return false;
+        const auto& isMine = pwallet->IsMine(parentOut);
+        if (isMine != ISMINE_SPENDABLE)
+        {
+            // If the wallet is hdwatchonly, check if it is a key we generated
+            if (!pwallet->IsHDWatchOnly() || isMine != ISMINE_WATCH_SOLVABLE)
+                return false;
+            const auto& meta = pwallet->mapKeyMetadata;
+            auto it = meta.find(CScriptID(parentOut.scriptPubKey));
+            if (it == meta.end() ||
+                it->second.hdKeypath.empty())
```

Comment:
Since we're not even required to be backwards compatible, I think adding a new piece of key metadata `fIsHDWatchOnly`, and adding that during `importaddress`(or some new call) when all keys are hdwatchonly keys(or a mixture of watchonly and stored private keys?).